pilot/pkg/model/push_context.go

```diff
@@ -78,9 +80,127 @@ type PushContext struct {
+// It is possible to split the endpoint tracking in future.
+type XDSUpdater interface {
+
+	// EDSUpdate is called when the list of endpoints or labels in a ServiceEntry is
+	// changed. For each cluster and hostname, the full list of active endpoints (including empty list)
+	// must be sent. The shard name is used as a key - current implementation is using the registry
+	// name.
+	EDSUpdate(shard, hostname string, entry []*IstioEndpoint) error
+
+	// SvcUpdate is called when a service port mapping definition is updated.
+	// This interface is WIP - labels, annotations and other changes to service may be
+	// updated to force a EDS and CDS recomputation and incremental push, as it doesn't affect
+	// LDS/RDS.
+	SvcUpdate(shard, hostname string, ports map[string]uint32, rports map[uint32]string)
+
+	// WorkloadUpdate is called by a registry when the labels or annotations on a workload have changed.
+	// The 'id' is the IP address of the pod for k8s if the pod is in the main/default network.
+	// In future it will include the 'network id' for pods in a different network, behind a zvpn gate.
+	// The IP is used because K8S Endpoints object associated with a Service only include the IP.
+	// We use Endpoints to track the membership to a service and readiness.
+	WorkloadUpdate(id string, labels map[string]string, annotations map[string]string)
+}
+
+// IstioEndpoint has the information about a single address+port for a specific
+// service and shard. Unlike ServiceInstance, it has less info and require less
+// allocation.
+type IstioEndpoint struct {
+
+	// Labels points to the workload or deployment labels.
+	Labels map[string]string
+
+	// Family indicates what type of endpoint, such as TCP or Unix Domain Socket.
+	// Default is TCP.
+	Family AddressFamily
+
+	// Address is the address of the endpoint, using envoy proto.
+	Address string
+
+	// EndpointPort is the port where the workload is listening, can be different
+	// from the service port.
+	EndpointPort uint32
+
+	// ServicePortName tracks the name of the port, to avoid 'eventual consistency' issues.
+	// Sometimes the Endpoint is visible before Service - so looking up the port number would
+	// fail. Instead the mapping to number is made when the clusters are computed. The lazy
+	// computation will also help with 'on-demand' and 'split horizon' - where it will be skipped
+	// for not used clusters or endpoints behind a gate.
+	ServicePortName string
+
+	// UID identifies the workload, for telemetry purpose.
+	UID string
+
+	// EnvoyEndpoint is a cached LbEndpoint, converted from the data, to
+	// avoid recomputation
+	EnvoyEndpoint *endpoint.LbEndpoint
+
+	// ServiceAccount holds the associated service account.
+	// May be replaced once a better optimization/struct is available for secure naming.
+	ServiceAccount string
+}
+
+// ServiceShards holds the set of endpoint shards of a service. Registries update
```

Comment:
Rename to ServiceInstanceDirectory, or ServiceInstanceIndex.

Per the operating model we defined elsewhere (for a consistent terminology), a service has one or more instances. And instance has one endpoint (which is a wasteful mapping). So, we are better off saying a Service has one or more network addressible instances. There is too much usage of the word endpoint here, that its hard to delineate which is k8s endpoint and which is envoy info.

Reply:
I don't see what is inefficient about this structure (https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L337) and https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L274.. 
The istio endpoint thing you created is in fact equally big (concatenation of both!) https://github.com/istio/istio/pull/8722/files#diff-ee39cd332931ab6f8c10713ce3446d09R134 ..

So the question is, if this is essentially replacing functionality, why create more confusion and messy code by adding entirely new structures which are almost replicas of the old ones, instead of literally optimizing the existing ones? The code currently is a sphagetti. Arbitrary environment variables, dependencies across packages that directly push, etc. Adding more of these data structures is only going to make things worse. 

Either way, ServiceShard makes absolutely no sense.. If its a service shard, its a group of services. Not an arbitrary grouping of endpoints in a service. And then there is EndpointShard to make matters worse. I can happily lgtm the code but any thing that breaks will end up coming to you :), as I don't understand the flow. Looking at comments from others, they seem to be in the same boat. So, its a question of maintainable code vs ad-hoc restructuring/duplication. I leave it up to you..

Reply:
Happy to add more documentation. I can rename "ServiceShards" to "ShardsOfService", and 
EndpointShard to "ServiceShard". Or other names - as long as they don't involve 'network', but the main idea is to break apart the list.

In master I will move some of those structs into v2 - most of the mess is caused by the remaining v1 discovery.go. I tried to minimize the PR and risks for 1.0 - which implies keeping existing code as a fail-safe and minimizing disruption. 

If you can optimize existing structs - without breaking existing code and in a reasonable time - please go for it. We have a requirement for big changes to be protected - so we need a flag to 
revert to old behavior and we need to keep the old structs for at least one version. After that NetworkEndpoint and old code can be removed, so it won't be a spaghetti. 

For example: NetworkEndpoint has both port name and int, but the Endpoint event only has name, and sometimes the Service object is received after Endpoints. As the code moves to incremental and 'eventually consistent' we need to deal with this - there are probably subtle bugs already in the code, in stress testing there are cases where the pod or service are not visible when the endpoint is. Some of the problem are hidden because everything is slow - but with
this PR EDS becomes very fast, and the inconsistencies visible.

And I agree that sharding is more complicated than keeping it all in a linear list, and doing O(n) operations on each update.